File: backend/src/Ahdah.Application/Expenses/Contracts/ExpenseRequests.cs
using System.ComponentModel.DataAnnotations;
using Ahdah.Application.Access.Contracts;

namespace Ahdah.Application.Expenses.Contracts;

public sealed record ExpenseCategoryQuery : PagedAccessQuery
{
    [StringLength(30)]
    public string? CategoryGroup { get; init; }

    [StringLength(20)]
    public string? ExpenseScope { get; init; }
}

public sealed record CreateExpenseCategoryRequest : IValidatableObject
{
    public Guid? ParentExpenseCategoryId { get; init; }

    [StringLength(30, MinimumLength = 2)]
    public string? CategoryCode { get; init; }

    [Required]
    [StringLength(150, MinimumLength = 1)]
    [RegularExpression("^(?=.*\\S)[^\\r\\n]+$")]
    public required string CategoryName { get; init; }

    [Required]
    [StringLength(30)]
    public required string CategoryGroup { get; init; }

    [Required]
    [StringLength(20)]
    public required string ExpenseScope { get; init; }

    [StringLength(500, MinimumLength = 1)]
    [RegularExpression("^(?=.*\\S)[^\\r\\n]+$")]
    public string? Description { get; init; }

    public bool RequiresSupplier { get; init; }
    public bool RequiresReceipt { get; init; } = true;
    public bool SupportsQuantityDetails { get; init; }

    [Range(0, int.MaxValue)]
    public int DisplayOrder { get; init; }

    public IEnumerable<ValidationResult> Validate(ValidationContext validationContext)
    {
        if (!ExpenseConstants.CategoryGroups.Contains(CategoryGroup))
        {
            yield return new ValidationResult("Category group is invalid.", [nameof(CategoryGroup)]);
        }

        if (!ExpenseConstants.CategoryScopes.Contains(ExpenseScope))
        {
            yield return new ValidationResult("Expense scope is invalid.", [nameof(ExpenseScope)]);
        }

        if (CategoryCode is { } code
            && (code.Trim().ToUpperInvariant() != code.Trim()
                || !System.Text.RegularExpressions.Regex.IsMatch(
                    code.Trim(), "^[A-Z0-9][A-Z0-9_-]{1,29}$")))
        {
            yield return new ValidationResult("Category code is invalid.", [nameof(CategoryCode)]);
        }
    }
}

public sealed record ExpenseAdvanceAllocationRequest
{
    public Guid UserAdvanceBalanceId { get; init; }
    public decimal Amount { get; init; }

    [StringLength(500, MinimumLength = 1)]
    [RegularExpression("^(?=.*\\S)[^\\r\\n]+$")]
    public string? Notes { get; init; }
}

public sealed record CreateExpenseRequest : IValidatableObject
{
    public Guid ExpenseCategoryId { get; init; }
    public Guid? ProjectId { get; init; }
    public DateOnly ExpenseDate { get; init; }
    public decimal Amount { get; init; }

    [Required]
    [StringLength(3, MinimumLength = 3)]
    [RegularExpression("^[A-Z]{3}$")]
    public required string CurrencyCode { get; init; }

    [Required]
    [StringLength(30)]
    public required string PaymentMode { get; init; }

    [Required]
    [StringLength(1000, MinimumLength = 1)]
    [RegularExpression("^(?=.*\\S)[^\\r\\n]+$")]
    public required string Description { get; init; }

    [StringLength(100, MinimumLength = 1)]
    [RegularExpression("^(?=.*\\S)[^\\r\\n]+$")]
    public string? InvoiceNumber { get; init; }

    [StringLength(100, MinimumLength = 1)]
    [RegularExpression("^(?=.*\\S)[^\\r\\n]+$")]
    public string? ReceiptNumber { get; init; }

    [StringLength(200, MinimumLength = 1)]
    [RegularExpression("^(?=.*\\S)[^\\r\\n]+$")]
    public string? MerchantName { get; init; }

    [StringLength(500, MinimumLength = 1)]
    [RegularExpression("^(?=.*\\S)[^\\r\\n]+$")]
    public string? ExpenseLocation { get; init; }

    [StringLength(1000, MinimumLength = 1)]
    [RegularExpression("^(?=.*\\S)[^\\r\\n]+$")]
    public string? Notes { get; init; }

    [MaxLength(100)]
    public IReadOnlyList<ExpenseAdvanceAllocationRequest> AdvanceAllocations { get; init; } = [];

    public IEnumerable<ValidationResult> Validate(ValidationContext validationContext)
    {
        if (ExpenseCategoryId == Guid.Empty)
        {
            yield return new ValidationResult("Expense category ID is required.", [nameof(ExpenseCategoryId)]);
        }

        if (ExpenseDate == default)
        {
            yield return new ValidationResult("Expense date is required.", [nameof(ExpenseDate)]);
        }

        if (!ExpenseRules.IsValidMoney(Amount))
        {
            yield return new ValidationResult("Amount must be positive and fit NUMERIC(18,2).", [nameof(Amount)]);
        }

        if (!ExpenseRules.IsSupportedCreationPaymentMode(PaymentMode))
        {
            yield return new ValidationResult(
                "Only AdvanceBalance and PersonalFunds are supported for expense creation.",
                [nameof(PaymentMode)]);
        }

        if (PaymentMode == ExpenseConstants.AdvanceBalancePaymentMode
            && !ExpenseRules.AllocationsMatch(AdvanceAllocations, Amount))
        {
            yield return new ValidationResult(
                "Advance allocations must be unique, positive, and total the expense amount.",
                [nameof(AdvanceAllocations), nameof(Amount)]);
        }

        if (PaymentMode == ExpenseConstants.PersonalFundsPaymentMode && AdvanceAllocations.Count != 0)
        {
            yield return new ValidationResult(
                "Personal-funds expenses cannot include advance allocations.",
                [nameof(AdvanceAllocations)]);
        }
    }
}

public sealed record ExpenseQuery : PagedAccessQuery
{
    [StringLength(30)]
    public string? Status { get; init; }
    public Guid? CategoryId { get; init; }
    public Guid? ProjectId { get; init; }
    public Guid? IncurredByUserId { get; init; }
    public Guid? SubmittedByUserId { get; init; }

    [StringLength(30)]
    public string? PaymentMode { get; init; }

    [StringLength(ExpenseConstants.MaximumSearchLength, MinimumLength = 2)]
    [RegularExpression("^[^\\r\\n]+$")]
    public string? Reference { get; init; }
}

public sealed record ExpenseHistoryQuery : PagedAccessQuery;

public sealed record ExpenseSubresourceQuery : PagedAccessQuery;

public sealed record ApproveExpenseRequest
{
    [Range(1, int.MaxValue)]
    public int ExpectedVersion { get; init; }
}

public sealed record RejectExpenseRequest
{
    [Range(1, int.MaxValue)]
    public int ExpectedVersion { get; init; }

    [Required]
    [StringLength(1000, MinimumLength = 1)]
    [RegularExpression("^(?=.*\\S)[^\\r\\n]+$")]
    public required string Reason { get; init; }
}

public sealed record AddExpenseDocumentRequest : IValidatableObject
{
    [Required]
    [StringLength(30)]
    public required string DocumentType { get; init; }

    [StringLength(100, MinimumLength = 1)]
    [RegularExpression("^(?=.*\\S)[^\\r\\n]+$")]
    public string? DocumentNumber { get; init; }

    public DateOnly? DocumentDate { get; init; }

    [StringLength(200, MinimumLength = 1)]
    [RegularExpression("^(?=.*\\S)[^\\r\\n]+$")]
    public string? IssuerName { get; init; }

    [Required]
    [StringLength(255, MinimumLength = 1)]
    [RegularExpression("^(?=.*\\S)[^\\r\\n/\\\\]+$")]
    public required string OriginalFileName { get; init; }

    [Required]
    [StringLength(1000, MinimumLength = 1)]
    public required string FileUrl { get; init; }

    [Required]
    [StringLength(150, MinimumLength = 3)]
    public required string MimeType { get; init; }

    [Range(1, long.MaxValue)]
    public long FileSizeBytes { get; init; }

    [Required]
    [StringLength(64, MinimumLength = 64)]
    [RegularExpression("^[A-Fa-f0-9]{64}$")]
    public required string Sha256Hash { get; init; }

    [Required]
    [StringLength(30)]
    public required string CaptureSource { get; init; }

    public bool IsPrimary { get; init; }

    [StringLength(1000, MinimumLength = 1)]
    [RegularExpression("^(?=.*\\S)[^\\r\\n]+$")]
    public string? Notes { get; init; }

    public IEnumerable<ValidationResult> Validate(ValidationContext validationContext)
    {
        if (!ExpenseConstants.DocumentTypes.Contains(DocumentType))
        {
            yield return new ValidationResult("Document type is invalid.", [nameof(DocumentType)]);
        }

        if (!ExpenseConstants.CaptureSources.Contains(CaptureSource))
        {
            yield return new ValidationResult("Capture source is invalid.", [nameof(CaptureSource)]);
        }

        if (!FileUrl.StartsWith("/", StringComparison.Ordinal)
            || FileUrl.Contains("..", StringComparison.Ordinal)
            || FileUrl.Contains("://", StringComparison.Ordinal))
        {
            yield return new ValidationResult(
                "File URL must be a safe application-relative path.",
                [nameof(FileUrl)]);
        }
    }
}

public sealed record ReimbursementQuery : PagedAccessQuery
{
    [StringLength(30)]
    public string? Status { get; init; }
    public Guid? ClaimantUserId { get; init; }
}

File: frontend/ahdah_app/lib/features/expenses/domain/expense_requests.dart
import 'dart:convert';

final class ExpenseAdvanceAllocationInput {
  const ExpenseAdvanceAllocationInput({
    required this.userAdvanceBalanceId,
    required this.amount,
    this.notes,
  });

  final String userAdvanceBalanceId;
  final String amount;
  final String? notes;
}

final class CreateExpenseInput {
  const CreateExpenseInput({
    required this.expenseCategoryId,
    required this.expenseDate,
    required this.amount,
    required this.currencyCode,
    required this.paymentMode,
    required this.description,
    this.projectId,
    this.invoiceNumber,
    this.receiptNumber,
    this.merchantName,
    this.expenseLocation,
    this.notes,
    this.advanceAllocations = const [],
  });

  final String expenseCategoryId;
  final String? projectId;
  final DateTime expenseDate;
  final String amount;
  final String currencyCode;
  final String paymentMode;
  final String description;
  final String? invoiceNumber;
  final String? receiptNumber;
  final String? merchantName;
  final String? expenseLocation;
  final String? notes;
  final List<ExpenseAdvanceAllocationInput> advanceAllocations;

  String get payloadFingerprint => toJsonBody();

  String toJsonBody() {
    const amountMarker = '__AHDAH_EXPENSE_AMOUNT__';
    final allocationMarkers = List.generate(
      advanceAllocations.length,
      (index) =>
          '__AHDAH_EXPENSE_ALLOCATION_$index'
          '__',
    );
    final encoded = jsonEncode({
      'expenseCategoryId': expenseCategoryId,
      'projectId': _optional(projectId),
      'expenseDate': _date(expenseDate),
      'amount': amountMarker,
      'currencyCode': currencyCode.trim().toUpperCase(),
      'paymentMode': paymentMode,
      'description': description.trim(),
      'invoiceNumber': _optional(invoiceNumber),
      'receiptNumber': _optional(receiptNumber),
      'merchantName': _optional(merchantName),
      'expenseLocation': _optional(expenseLocation),
      'notes': _optional(notes),
      'advanceAllocations': [
        for (var index = 0; index < advanceAllocations.length; index++)
          {
            'userAdvanceBalanceId':
                advanceAllocations[index].userAdvanceBalanceId,
            'amount': allocationMarkers[index],
            'notes': _optional(advanceAllocations[index].notes),
          },
      ],
    });
    var result = encoded.replaceFirst('"$amountMarker"', amount);
    for (var index = 0; index < allocationMarkers.length; index++) {
      result = result.replaceFirst(
        '"${allocationMarkers[index]}"',
        advanceAllocations[index].amount,
      );
    }
    return result;
  }
}

final class CreateExpenseCategoryInput {
  const CreateExpenseCategoryInput({
    required this.categoryName,
    required this.categoryGroup,
    required this.expenseScope,
    this.parentExpenseCategoryId,
    this.categoryCode,
    this.description,
    this.requiresSupplier = false,
    this.requiresReceipt = true,
    this.supportsQuantityDetails = false,
    this.displayOrder = 0,
  });

  final String? parentExpenseCategoryId;
  final String? categoryCode;
  final String categoryName;
  final String categoryGroup;
  final String expenseScope;
  final String? description;
  final bool requiresSupplier;
  final bool requiresReceipt;
  final bool supportsQuantityDetails;
  final int displayOrder;

  Map<String, Object?> toJson() => {
    'parentExpenseCategoryId': _optional(parentExpenseCategoryId),
    'categoryCode': _optional(categoryCode)?.toUpperCase(),
    'categoryName': categoryName.trim(),
    'categoryGroup': categoryGroup,
    'expenseScope': expenseScope,
    'description': _optional(description),
    'requiresSupplier': requiresSupplier,
    'requiresReceipt': requiresReceipt,
    'supportsQuantityDetails': supportsQuantityDetails,
    'displayOrder': displayOrder,
  };
}

final class ApproveExpenseInput {
  const ApproveExpenseInput({required this.expectedVersion});
  final int expectedVersion;
  String get payloadFingerprint => jsonEncode(toJson());
  Map<String, Object?> toJson() => {'expectedVersion': expectedVersion};
}

final class RejectExpenseInput {
  const RejectExpenseInput({
    required this.expectedVersion,
    required this.reason,
  });

  final int expectedVersion;
  final String reason;
  String get payloadFingerprint => jsonEncode(toJson());
  Map<String, Object?> toJson() => {
    'expectedVersion': expectedVersion,
    'reason': reason.trim(),
  };
}

String _date(DateTime value) =>
    '${value.year.toString().padLeft(4, '0')}-'
    '${value.month.toString().padLeft(2, '0')}-'
    '${value.day.toString().padLeft(2, '0')}';

String? _optional(String? value) {
  final normalized = value?.trim();
  return normalized == null || normalized.isEmpty ? null : normalized;
}
